pkg/backup/item_backupper.go

```diff
@@ -253,11 +253,34 @@ func (ib *defaultItemBackupper) backupItem(logger logrus.FieldLogger, obj runtim
 		return false, kubeerrs.NewAggregate(backupErrs)
 	}
 
+	// group version of this object
+	// Used on filepath to backup up all groups and versions
+	version := resourceVersion(obj)
+
+	// Getting the preferred group version of this resource
+	gvr_preferred, _, err := ib.discoveryHelper.ResourceFor(groupResource.WithVersion(""))
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+	versionpreferred := gvr_preferred.Version
+
 	var filePath string
-	if namespace != "" {
-		filePath = filepath.Join(api.ResourcesDir, groupResource.String(), api.NamespaceScopedDir, namespace, name+".json")
+
+	// API Group version is now part of path of backup as a subdirectory
+	// it will add a prefix to subdirectory name for the preferred version
+	if version == versionpreferred {
```

Comment:
What if we re-wrote this slightly, to reduce duplicating the branching on filepath?

Something like

```go
versionPath := version
if version == versionpreferred {
versionPath = version+api.PreferredVersionDir
}
		if namespace != "" {
			filePath = filepath.Join(api.ResourcesDir, groupResource.String(), versionPath, api.NamespaceScopedDir, namespace, name+".json")
		} else {
			filePath = filepath.Join(api.ResourcesDir, groupResource.String(), versionPath, api.ClusterScopedDir, name+".json")
		}
```

That way we can keep 2 assignments for `filePath` like before, and still manipulate the form of the version directory.

Reply:
I added the piece of code as suggested with an amended commit. I ran the regular tests using a slightest enhanced version of the [script](https://github.com/brito-rafa/velero-test/blob/master/2251-dev-test.sh) with the following [output ](https://github.com/brito-rafa/velero-test/blob/master/testing-2251-20200405.log). The test was taking a backup with the patched version on a 1.16 cluster and restoring using standard Velero 1.3.1 on a K8s 1.18.

Reply:
Flagging as resolved, feel free to add more.